### PR TITLE
fix(variants): scope HPA/ScaledObject list calls to tracked namespaces

### LINKED ISSUE #1134: perf: scope annotationSourcedVariants List calls to tracked namespaces
## Problem

`annotationSourcedVariants` in `internal/utils/variant.go` issues two unfiltered `List` calls on every engine tick:

```go
k8sClient.List(ctx, &hpaList)          // fetches all HPAs cluster-wide
k8sClient.List(ctx, &soList)           // fetches all ScaledObjects cluster-wide
```

These are cache-local reads (no API round-trip via the controller-runtime informer cache), but in large clusters with many unrelated HPAs and ScaledObjects they waste CPU iterating and filtering objects that will never carry `llm-d.ai/managed: "true"`.

Identified in PR #1123 review comment: https://github.com/llm-d/llm-d-workload-variant-autoscaler/pull/1123#issuecomment-4433913451

## Proposed Fix

Scope the lists to namespaces already known to contain managed objects, using `Datastore.ListTrackedNamespaces()` (already exists on the interface). Issue `client.InNamespace(ns)` per tracked namespace and union results. Fall back to cluster-wide when the tracked-namespaces list is empty (first tick before any watch event fires).

This requires threading `trackedNamespaces []string` through the call chain:

1. `annotationSourcedVariants(ctx, client, trackedNamespaces []string)`
2. `readyVariantAutoscalings(ctx, client, trackedNamespaces []string)`
3. `filterVariantsByScaleTargetAccessor(ctx, client, trackedNamespaces []string, ...)`
4. `ActiveVariantAutoscaling(ctx, client, trackedNamespaces []string)` — public API change
5. `InactiveVariantAutoscaling(ctx, client, trackedNamespaces []string)` — same
6. Engine call site (`engine.go:259`) — pass `e.ds.ListTrackedNamespaces()`, requires adding `ds datastore.Datastore` to the `Engine` struct

## Acceptance Criteria

- [ ] `annotationSourcedVariants` lists only per-namespace when tracked namespaces are known
- [ ] Falls back to cluster-wide on empty tracked-namespaces list
- [ ] `make test` passes

### PR DESCRIPTION
**Pull request by Goutham-Annem**
## Summary

Fixes #1134.

`annotationSourcedVariants()` issued two unscoped cluster-wide `List` calls (HPAs and ScaledObjects) on every engine tick, reading the entire cluster cache even when only a small number of namespaces were actually tracked.

**Root cause**: `k8sClient.List(ctx, &hpaList)` and `k8sClient.List(ctx, &soList)` had no `client.ListOption` to restrict the query, as noted by the `TODO(#1134)` comments in the code.

**Fix**: Add a `trackedNamespaces []string` parameter through the call chain. When non-empty, perform one `client.InNamespace(ns)` scoped call per tracked namespace and union the results. When empty/nil, fall back to the original cluster-wide behaviour.

The parameter is threaded through:
- `annotationSourcedVariants(ctx, k8sClient, trackedNamespaces)`
- `readyVariantAutoscalings(ctx, k8sClient, trackedNamespaces)`
- `filterVariantsByScaleTargetAccessor(ctx, client, filter, filterName, trackedNamespaces)`
- `ActiveVariantAutoscaling(ctx, client, trackedNamespaces)`
- `InactiveVariantAutoscaling(ctx, client, trackedNamespaces)`
- `ActiveVariantAutoscalingByModel(ctx, client, trackedNamespaces)`
- `InactiveVariantAutoscalingByModel(ctx, client, trackedNamespaces)`

The scalefromzero engine already holds `Datastore` and now passes `e.Datastore.ListTrackedNamespaces()`. The saturation engine gains an exported `Datastore datastore.Datastore` field (zero-value safe — `nil` means cluster-wide) that is wired from `cmd/main.go` via `engine.Datastore = ds`.

## Changes

- `internal/utils/variant.go`: add `namespacedListCalls` helper; thread `trackedNamespaces` through the full call chain; remove TODO(#1134) comments
- `internal/engines/saturation/engine.go`: add `Datastore datastore.Datastore` field; pass `ListTrackedNamespaces()` to `ActiveVariantAutoscaling`
- `internal/engines/scalefromzero/engine.go`: pass `ListTrackedNamespaces()` to `InactiveVariantAutoscaling`
- `cmd/main.go`: wire `ds` into saturation engine via `engine.Datastore = ds`
- `internal/utils/variant_test.go`: add `TestAnnotationSourcedVariantsNamespaceScoping` (3 cases: single namespace, multi-namespace union, nil fallback); update all call sites to pass `nil`
- `internal/engines/scalefromzero/engine_test.go`: update call site to pass `nil`

## Test plan

- [x] New `TestAnnotationSourcedVariantsNamespaceScoping` tests pass
- [x] All existing unit tests pass (`go test ./internal/utils/... ./internal/engines/...`)
- [x] Full unit test suite passes (`go test ./...`, excluding pre-existing e2e/chart failures requiring a live cluster)
- [x] `go build ./...` clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by lionelvillard** [comment]
@Goutham-Annem can you please fix the linter error?

File: internal/utils/variant.go [hunk 1]
```diff
@@ -194,53 +199,64 @@ func readyVariantAutoscalings(ctx context.Context, k8sClient client.Client) []wv
...
-		return nil, fmt.Errorf("listing HPAs: %w", err)
-	}
-	for i := range hpaList.Items {
-		hpa := &hpaList.Items[i]
-		if !annotations.IsManaged(hpa) || !hpa.DeletionTimestamp.IsZero() {
-			continue
+	for _, opts := range nsCalls {
+		var hpaList autoscalingv2.HorizontalPodAutoscalerList
+		if err := k8sClient.List(ctx, &hpaList, opts...); err != nil {
+			return nil, fmt.Errorf("listing HPAs: %w", err)
```
Comment (ev-shindin) [inline internal/utils/variant.go:230]: Inside the per-namespace HPA loop, a `List` error on *any* namespace does `return nil, err`, discarding every variant already collected from earlier namespaces this call. The ScaledObject loop below (`L253–267`) instead keeps partial results on error. With one cluster-wide call this asymmetry didn't exist (no prior data to lose); the per-namespace loop introduces a new partial-failure mode where one namespace's transient error wipes the whole tick's HPA discovery (it's non-fatal at the caller, so the tick just sees zero HPAs). Suggest matching the ScaledObject pattern — record the error, continue, and return partial results.

File: internal/engines/saturation/engine.go [hunk 2]
```diff
@@ -359,7 +364,11 @@ func (e *Engine) optimize(ctx context.Context) (retErr error) {
 		logger.Info("Scaling to zero is enabled")
 	}
 
-	activeVAs, _, err := utils.ActiveVariantAutoscaling(ctx, e.client)
+	var trackedNamespaces []string
+	if e.Datastore != nil {
```
Comment (ev-shindin) [inline internal/engines/saturation/engine.go:459]: The new `Datastore` scoping wiring (`engine.Datastore = ds` in `cmd/main.go`, the `e.Datastore != nil` branch here) has no saturation-package test - nothing sets `Engine.Datastore` or checks the nil-fallback. A dropped assignment or inverted nil-check would pass the whole suite. Add a test seeding a `datastore.Datastore` with tracked namespaces + managed scalers in tracked and untracked namespaces, asserting only tracked ones are optimized (plus a `Datastore == nil` -> cluster-wide case).

File: internal/engines/scalefromzero/engine.go [hunk 3]
```diff
@@ -127,7 +127,7 @@ func (e *Engine) optimize(ctx context.Context) error {
 	logger := log.FromContext(ctx)
 
 	// Get all inactive (replicas == 0) VAs
-	inactiveVAs, scaleTargets, err := utils.InactiveVariantAutoscaling(ctx, e.client)
+	inactiveVAs, scaleTargets, err := utils.InactiveVariantAutoscaling(ctx, e.client, e.Datastore.ListTrackedNamespaces())
```
Comment (ev-shindin) [inline internal/engines/scalefromzero/engine.go:132]: Similar to `saturation/engine.go` : the only touched test just updates the call signature with `nil`; no engine-level test asserts scoping to tracked namespaces.

File: internal/utils/variant.go (hunk 1 above)
Comment (ev-shindin) [inline internal/utils/variant.go:230]: Two untested new paths: `namespacedListCalls` isn't unit-tested directly for the empty/single/multi cases (only covered indirectly), and the multi-namespace list-error path (succeed on ns1, fail on ns2) -which would surface the HPA data-loss above - has no test. Add an `interceptor.Funcs` case that fails one namespace's List.

**Review by ev-shindin — COMMENTED** [review]
LGTM. Thanks @Goutham-Annem ! Please address my comments

**Comment by asm582** [comment]
FYI, this PR looks like a duplicate of https://github.com/llm-d/llm-d-workload-variant-autoscaler/pull/1189
